### ISSUE TITLE
[BUG] auth doesnt work, especially keycloak

### ISSUE THREAD

**Issue by Ev0genii** [Lissy93/dashy#1290]
### Environment

Self-Hosted (Docker)

### System

debian

### Version

2.1.1

### Describe the problem

Deploy dashy in kubernetes with ingress
tried to make auth by dashy documentation, but it didnt work, even basic auth with user hash and title
also auth parameters as enableGuestAccess doesnt work
i think its something in last update, that interfere auth work
last git update was a week ago, maybe u didnt update documentation

### Please tick the boxes

- [X] You have explained the issue clearly, and included all relevant info
- [X] You are using a [supported](https://github.com/Lissy93/dashy/blob/master/.github/SECURITY.md#supported-versions) version of Dashy
- [X] You've checked that this [issue hasn't already been raised](https://github.com/Lissy93/dashy/issues?q=is%3Aissue)
- [X] You've checked the [docs](https://github.com/Lissy93/dashy/tree/master/docs#readme) and [troubleshooting](https://github.com/Lissy93/dashy/blob/master/docs/troubleshooting.md#troubleshooting) guide ![important](https://img.shields.io/badge/Important!-F6094E)
- [X] You agree to the [code of conduct](https://github.com/Lissy93/dashy/blob/master/.github/CODE_OF_CONDUCT.md#contributor-covenant-code-of-conduct)

**Comment by juancarlosm** [comment]
Same here. Basic auth is not working.

**Comment by randyjc** [comment]
Was about to give up, because I couldn't find what I was doing wrong.
Tried a couple of different options, but below is latest.
After applying below to the config, I also get the message that I'm not logged in as admin.
When trying to force via URL to login I just get redirected to Dashy homepage.

https://url/login redirects to https://url/

I deleted cache, restarts, incognito, other browsers etc.

```yaml
appConfig:
  theme: nord-frost
  layout: auto
  iconSize: medium
  language: en
  auth:
    users:
    - user: randyjc
      hash: <REDACTED>
      type: admin
pageInfo:
  title: Dashy
  description: Welcome to your new dashboard!
  navLinks:
    - title: GitHub
      path: https://github.com/Lissy93/dashy
    - title: Documentation
      path: https://dashy.to/docs
sections:
  - name: Getting Started
    icon: fas fa-rocket
    items:
      - title: Dashy Live
        description: Development a project management links for Dashy
        icon: https://i.ibb.co/qWWpD0v/astro-dab-128.png
        url: https://live.dashy.to/
        target: newtab
        id: 0_1481_dashylive
      - title: GitHub
        description: Source Code, Issues and Pull Requests
        url: https://github.com/lissy93/dashy
        icon: favicon
        id: 1_1481_github
```

V2.1.1 via PikaPods

**Comment by artemdanielov** [comment]
same here, auth (basic) soesnt work =(

**Comment by ajbrigg49** [comment]
I am new to Dashy (Just set it up about a week ago.) Today I was following the basic auth docs and I am also seeing the same issue. Opening Dashy, I just get to my dashboard. If I go to config, update config, edit config, I see a message across the bottom "You cannot write changes to disk because you are not logged in as an administrator". So it would seem it knows about my new user account that is an admin, however it is not redirecting to login. I've tried multiple browsers (Edge, chromium, Chrome), and multiple PCs, but there is no change. If it matters, I added the auth config by hand from ssh use VI. 

Thanks

Andy

**Comment by coldshouldermedia** [comment]
I seem to be having a similar issue.

**Comment by chrgraham** [comment]
+1 Basic auth is NOT working, new user running Docker, auth is configured, dashboard is completely exposed and unsecured. 

Auth tickets should be a priority.

**Comment by Pytonballoon810** [comment]
Is this still going on? I think I have the same problem but cant seem to find a fix if there is maybe someone can direct me into the right direction?

**Comment by coldshouldermedia** [comment]
I've not found a solution yet.  So for now I've just disabled Dashy.

**Comment by IbaiBuR** [comment]
I am experiencing the same issue too... Neither keycloak nor basic auth working... 

Lets hope @Lissy93 can fix it soon somehow.

**Comment by bloodysneaker** [comment]
+1
No login page showing after enabling auth.
With the aforementioned work-around of "enableguestaccess" I've the login button but even if I authenticate with an Admin user, I'm no able to make any conf changes.

**Comment by Maretzky85** [comment]
+1 on docker swarm
2.1.0 docker image works fine tho.
Keeping previous version until fix is available :)

**Comment by nOw-Ay** [comment]
+1, I have exactly the same problem that is described here and in the other issues (logout button without even logging-in and keycloak weird errors).
Reverting to 2.1.0

**Comment by coderunway** [comment]
+1 , This is not working on Dashy 2.1.1 too. Does not work even after adding **enableGuestAccess: true** and **disableConfigurationForNonAdmin: true** parameters. I have tried it in the incognito mode browser too, to ensure its not a caching issue. Any advice is appreciated.

```
appConfig:
    auth:
        enableGuestAccess: false 
        users:
          - user: user1
            hash: ############
            type: admin
          - user: user2
            hash: ##############
            type: normal
      disableConfigurationForNonAdmin: true
```

**Comment by jakob1379** [comment]
Keycloak works perfectly fine for me.

The needing element is an attribute mapper providing roles and groups to simple attributes "roles" and "groups" in the access token. 

![image](https://github.com/Lissy93/dashy/assets/17257805/f851fb36-a70c-4aa7-8943-18266829c1f7)

Does not work in conjunction with basic auth though

**Comment by djessich** [comment]
Any progress on this?
For me, it also does not work with 2.1.1. With 2.1.0 everything works perfectly though...

**Comment by jakob1379** [comment]
> Any progress on this? For me, it also does not work with 2.1.1. With 2.1.0 everything works perfectly though...

Have you made sure you get the groups with your requests? It works when you add them to the access token

**Comment by TheRealGramdalf** [comment]
I believe there are several reasons why this could be happening - in my case it was a combination of two issues.

- Problem 1: User Error
    - Contrary to what the documentation says, I decided to bind mount `./config:/app/public`. This was a mistake, since `/app/public` is pre-populated with a sample configuration and some other files - including the website icon etc. **Only bind mount individual files into `/app/public`!!**
    - _Note: In  the [release notes for 2.1.1](https://github.com/Lissy93/dashy/releases/tag/2.1.1), there was mention about a configuration overhaul - I would wait on this to be released before making any requests/suggestions.
- Problem 2: First build
    - I haven't had a chance to look fully into this yet, but I believe it's due to how the dashy docker image is built. It runs a node.js server (which is required to support functions like config editing and more), and a "static" site generated from the config. The static site is built at runtime with `yarn`, and can be manually started with `yarn build` (or through the UI via `Update Configuration > Rebuild Application`). The generated site is placed at `/app/dist/`, which is non-persistent - if you recreate the container (which happens when running `docker compose down`, or `docker compose up -d` with an already running container), the contents of `/app/dist` will be lost. I believe that upon first startup, the site is run directly from source while the first build is created - which leads to inconsistencies in login mentioned in this issue. If (like me) there are issues with auto rebuild, the static site won't be generated automatically - you need to trigger it manually via any of the above methods. Once built, login functions as expected.
    - @Lissy93 do you mind giving this a quick read and response? A yes/no would suffice, I know you are busy with v2.1.2/v3 among other things.

As for solutions to the problem, make sure the static component is built upon container recreation. Since the release notes mentioned a big overhaul, I won't do anything right now, but I have an idea as for how to fix the issue at it's root, should it still exist - make `/app/dist` persist to an extent (or require `yarn build` to complete before the container becomes accessible). If a maintainer can give me the go ahead (i.e. this won't be solved by anything that's currently in the works), I'll draft up a PR with the fix (basically add documentation for an additional bind mount, `./cache:/app/dist` or the like, and make a few changes to make it work (there's an RM command somewhere that needs a change)).

**Comment by TheRealGramdalf** [comment]
_These mentions are just for letting people involved in the other bugs know, sorry for the spam_

**Comment by zell-mbc** [comment]
Thanks for looking int this @TheRealGramdalf.
I tried at my end again, unfortunately I couldn't get things to work.
- Added the Keycloack mapper as per the screenshot of @jakob1379, no change unfortunately
- Ran` sudo docker exec -it dashboard yarn` as per your note above and still won't get a prompt to logon

Back to v2.1.0 and things are working again…

**Comment by TheRealGramdalf** [comment]
>     * Ran` sudo docker exec -it dashboard yarn` as per your note above and still won't get a prompt to logon

Did you do  `sudo docker exec -it dashboard yarn` or `sudo docker exec -it dashboard yarn build`? The latter is what you need for it to build, on my server it takes about ~100s. Make sure to clear all cache (especially if you have the service worker enabled) for it to take effect - there may be residual login cookies etc that mess with things.

**Comment by okietech** [comment]
I just tried your fix and it worked like a charm.  Thanks!

**Comment by zell-mbc** [comment]
> Did you do `sudo docker exec -it dashboard yarn` or `sudo docker exec -it dashboard yarn build`? 

The first one :-)
Just tried again with yarn build…and bingo! Got my logon dialog!
I think that's a workable solution for me until this get fixed for good.

Thanks a lot for helping along, really appreciated!

**Comment by Unacoli** [comment]
Even with version 2.1.0, I don't achieve keycloak auth to work :( It gives a never ending refresh on the dashy loading screen. Did someone have a working configuration with keycloak to provide ?

**Comment by DimitrijaGj** [comment]
Hello,

I'm encountering a couple of issues with Dashy after deploying it using Docker Compose. Here's a summary of my steps and the problems I'm facing:

Initial Deployment: I deployed Dashy using the command docker compose up -d. After deployment, the login screen did not appear as expected.

Rebuilding Dashy: To troubleshoot, I rebuilt Dashy using two methods:

Through the Dashy UI by selecting "Rebuild Application."
Via command line with sudo docker exec -it <container name> yarn build.
After rebuilding, the login screen became available, resolving the initial issue.

Configuration Update Problem: However, a new problem arose where updates made to the Dashy configuration through the UI are not being saved. Interestingly, the changes appear in the configuration file when viewed with cat or vim, but these updates don't reflect within the application itself when I reload the browser, tried also in incognito mode clear cookis.
Specifically, I'm looking for help with the following:

Why are the configuration changes not persisting in the UI, even though they're visible in the configuration file?
Is there a step I'm missing or a known issue that prevents configuration updates from being saved properly when using Docker?
Any insights, suggestions, or guidance on how to resolve this issue would be greatly appreciated. Thank you!

**Comment by TheRealGramdalf** [comment]
Configuration changes require a rebuild before they take effect. Just trigger that after making changes to the config and you should be good.

**Comment by DimitrijaGj** [comment]
> Configuration changes require a rebuild before they take effect. Just trigger that after making changes to the config and you should be good.

Thanks for the fast response. However, the problem I face after rebuild is that I can not change the configuration in UI, so to say interactively. I do the changes, like add section, as can be seen on the screenshots. I added the "test section" clicked on save disk, after refreshing the page "test section" is gone, but is, in fact, in the config.yml file. 
It is very inconvenient to rebuild the app after every change. 
 So long story short, I have only problem that after I run it with docker compose up -d, anyone can see the dashboard after just deleting the /login in the URL IP:Port/login is there a workaround I can do to avoid that??
 

![Screenshot 2024-02-16 at 08 00 27](https://github.com/Lissy93/dashy/assets/98087281/f0d3a0dc-572b-485e-9b97-00ebff453ecd)
![Screenshot 2024-02-16 at 08 00 37](https://github.com/Lissy93/dashy/assets/98087281/e3e9aed1-a978-4b70-ba83-d5b5c21a1bd7)

**Comment by TheRealGramdalf** [comment]
Essentially the issue is that you **_must_** rebuild after making any changes, period. This used to happen automatically, but it was recently broken (as of 2.11 iirc?). I haven't looked into the specifics exactly, but that's the TL;DR

**Comment by DimitrijaGj** [comment]
> Essentially the issue is that you **_must_** rebuild after making any changes, period. This used to happen automatically, but it was recently broken (as of 2.11 iirc?). I haven't looked into the specifics exactly, but that's the TL;DR

Thx for the answer. I revert it to 2.1.0 and it is working fine. I will wait till the bug is fix in the next releases.

**Comment by jerry-f** [comment]
&gt; Even with version 2.1.0, I can't get keycloak authentication to work :( It provides a never-ending refresh on the dash loading screen. Does anyone have a working configuration for keycloak?

I also can't use keycloak for authentication.

**Comment by CrazyWolf13** [comment]
Hi
This problems seems to be related to dashy not automatically rebuilding on changes, see #1479 and https://github.com/Lissy93/dashy/issues/1290#issuecomment-1884249018 
A new release has just been published, which should fix this issue, so please upgrade to the latest version.
If this is not the case just respond back to this ticket and I'll reopen this issue.
Thanks and have a nice day.

**Comment by Unacoli** [comment]
Hello @CrazyWolf13 !

I am still having issue with dashy blocked in a never ending refresh. I have the page with auth with keycloak, and when I am login in, I have dashy that is stuck refreshing and not working.

I don't know if its still an issue with dashy or if it is a config problem :(

**Comment by CrazyWolf13** [comment]
@Unacoli 
Thanks for the feedback, can you confirm you are indeed running the latest version of dashy and can you try to repull and redeploy the container?

Also would you mind sharing your conf.yml ?

**Comment by zell-mbc** [comment]
I have a sort of working setup in case you want to compare notes.
"Sort of" because when I do a `docker compose down --rmi all && docker compose up -d` I have to do a separate rebuild via `docker exec -it dashboard yarn build`

Without the later docker command Dashy will work, but not ask for credentials, which is the opposite of fail-safe I guess :-)

**Comment by zell-mbc** [comment]
Yes, this is what I am running:

`dashboard  | Using Dashy V-2.1.2. Update Check Complete`

**Comment by Unacoli** [comment]
Yes I am indeed running the latest version.

`pageInfo:
  title: Hub STJV
sections:
  - name: Example Section
appConfig:
  theme: callisto
  language: fr
  layout: auto
  iconSize: medium
  startingView: default
  defaultOpeningMethod: newtab
  statusCheck: false
  statusCheckInterval: 0
  faviconApi: allesedv
  routingMode: history
  enableMultiTasking: false
  widgetsAlwaysUseProxy: false
  webSearch:
    disableWebSearch: false
    searchEngine: duckduckgo
    openingMethod: newtab
    searchBangs: {}
  enableFontAwesome: true
  enableMaterialDesignIcons: false
  hideComponents:
    hideHeading: false
    hideNav: false
    hideSearch: false
    hideSettings: false
    hideFooter: false
  auth:
    enableGuestAccess: false
    users: []
    enableKeycloak: true
    keycloak:
      serverUrl: [URL]
      realm: [Realm]
      clientId: [id]
      legacySupport: false
      idpHint: [idp]
  showSplashScreen: false
  preventWriteToDisk: false
  preventLocalSave: false
  disableConfiguration: false
  disableConfigurationForNonAdmin: false
  allowConfigEdit: true
  enableServiceWorker: false
  disableContextMenu: false
  disableUpdateChecks: false
  disableSmartSort: false
  enableErrorReporting: false
`

**Comment by Unacoli** [comment]
> I have a sort of working setup in case you want to compare notes. "Sort of" because when I do a `docker compose down --rmi all && docker compose up -d` I have to do a separate rebuild via `docker exec -it dashboard yarn build`
> 
> Without the later docker command Dashy will work, but not ask for credentials, which is the opposite of fail-safe I guess :-)

The command sadly didn't solved the refresh problem :( I really don't understand if the problem come from me or not since its gave me the right connexion prompt, and after login its locked in the refreshing state (i also see my connexion log on dashy on the keycloak log)

**Comment by Unacoli** [comment]
I've investigated a little more browsing tickets, and I find the answer to my problem !

In Keycloak, in my dashy client, I had to turn off this option : 
![image](https://github.com/Lissy93/dashy/assets/29428817/51201a43-fca6-4523-b3c2-bc04947f4f10)

It works perfectly now, thank you all !

**Comment by e-sacchi** [comment]
I also have this problem, normal login works, keycloak not, even with this feature disabled... No login page and config file open to anyone. Any updates on this?

**Comment by Kirkland-gh** [comment]
Keycloak is working for me now on 3.0.0 when I go to dashy.domain.com/login. Is anyone else able to test?

**Comment by kutekroma** [comment]
After the authorization window, enter the login and password. Next, the infinite cycle begins. Where can I see detailed logs?

**Comment by kutekroma** [comment]
Console output after keycloak authorization: 

Error while trying to use the following icon from the Manifest: https://***.domen.com/web-icons/dashy-pwa_144x144.png (Download error or resource isn't a valid image)

Warning : Deprecation warning: The `simple-icons` entrypoint will be removed in the next major. Please switch to using `import * as icons from "simple-icons/icons"` if you need an object with all the icons.

**Comment by kutekroma** [comment]
> Keycloak is working for me now on 3.0.0 when I go to dashy.domain.com/login. Is anyone else able to test?




> Keycloak теперь работает у меня в версии 3.0.0, когда я захожу на Dashy.domain.com/login. Кто-нибудь еще может протестировать?



> Keycloak is working for me now on 3.0.0 when I go to dashy.domain.com/login. Is anyone else able to test?



> Keycloak is working for me now on 3.0.0 when I go to dashy.domain.com/login. Is anyone else able to test?

Can you show conf.yml and a screenshot of the clients setting in keycloak?
Have you set up roles and groups separately in the client?

**Comment by kutekroma** [comment]
> Keycloak is working for me now on 3.0.0 when I go to dashy.domain.com/login. Is anyone else able to test?

can you show conf.yml and a screenshot of the clients setting in keycloak?
Have you set up roles and groups separately in the client?

**Comment by Kirkland-gh** [comment]
> > Keycloak is working for me now on 3.0.0 when I go to dashy.domain.com/login. Is anyone else able to test?
> 
> can you show conf.yml and a screenshot of the clients setting in keycloak? Have you set up roles and groups separately in the client?

Only one user so no roles or groups. 

```
  auth:
    enableGuestAccess: false
    users: []
    enableKeycloak: true
    keycloak:
      serverUrl: https://keycloak.example.com
      realm: example-realm
      clientId: example-clientid
      legacySupport: false
```

**Comment by Lissy93** [comment]
I hadn't realized that the infinite refresh loop was client-side, sorry misunderstood and thought you were talking about a Docker issue.

This should be fixed by db9d7e362dcd5d08688a6eb32521b0499f9ce809  in #1542 and released in [3.0.1](https://github.com/Lissy93/dashy/releases/tag/3.0.1)

**Comment by kutekroma** [comment]
> I hadn't realized that the infinite refresh loop was client-side, sorry misunderstood and thought you were talking about a Docker issue.
> 
> This should be fixed by [db9d7e3](https://github.com/Lissy93/dashy/commit/db9d7e362dcd5d08688a6eb32521b0499f9ce809) in #1542 and released in [3.0.1](https://github.com/Lissy93/dashy/releases/tag/3.0.1)

Updated to version 3.0.1 and getting an error
![2024-05-02 12_29_45-Dashy](https://github.com/Lissy93/dashy/assets/22984319/d70cf729-57cb-4a13-9cec-32c09a643c4b)

Moreover, keycloak authorization is successful and the session is active for the user.

My conf.yml

`pageInfo:
  title: Home
  description: ***
  navLinks:
    - title: Gitlab
      path: https://gitlab.***.com
    - title: Books
      path: https://books.***.com
appConfig:
  theme: colorful
  auth:
    enableGuestAccess: false
    enableKeycloak: true
    keycloak:
      serverUrl: https://keycloak.***.com
      realm: KC1
      clientId: Home
  layout: auto
  iconSize: medium
  language: ru
sections:
  - name: Example Section
    icon: far fa-rocket
    displayData:
        showForKeycloakUsers:
          roles: ['admin']
    items:
      - title: GitHub
        description: Dashy source code and docs
        icon: fab fa-github
        url: https://github.com/Lissy93/dashy
        id: 0_1473_github
      - title: Issues
        description: View open issues, or raise a new one
        icon: fas fa-bug
        url: https://github.com/Lissy93/dashy/issues
        id: 1_1473_issues`

**Comment by kutekroma** [comment]
Great! It works with version 3.0.1
THANKS!!!!!
